Firestore/core/src/firebase/firestore/core/transaction.mm

```diff
@@ -0,0 +1,209 @@
+    const DocumentKey& key) {
+  absl::optional<SnapshotVersion> version = GetVersion(key);
+
+  if (version.has_value() && version.value() == SnapshotVersion::None()) {
+    // The document to update doesn't exist, so fail the transaction.
+    return Status{FirestoreErrorCode::Aborted,
+                  "Can't update a document that doesn't exist."};
+  } else if (version.has_value()) {
+    // Document exists, just base precondition on document update time.
+    return Precondition::UpdateTime(version.value());
+  } else {
+    // Document was not read, so we just use the preconditions for a blind
+    // update.
+    return Precondition::Exists(true);
+  }
+}
+
+void Transaction::Set(const DocumentKey& key, ParsedSetData&& data) {
+  WriteMutations(std::move(data).ToMutations(key, CreatePrecondition(key)));
+}
+
+void Transaction::Update(const DocumentKey& key, ParsedUpdateData&& data) {
+  StatusOr<Precondition> maybe_precondition = CreateUpdatePrecondition(key);
+  if (!maybe_precondition.ok()) {
+    last_write_error_ = maybe_precondition.status();
+  } else {
+    WriteMutations(
+        std::move(data).ToMutations(key, maybe_precondition.ValueOrDie()));
+  }
+}
+
+void Transaction::Delete(const DocumentKey& key) {
+  FSTMutation* mutation =
+      [[FSTDeleteMutation alloc] initWithKey:key
+                                precondition:CreatePrecondition(key)];
+  WriteMutations({mutation});
+
+  // Since the delete will be applied before all following writes, we need to
+  // ensure that the precondition for the next write will be exists: false.
+  read_versions_[key] = SnapshotVersion::None();
+}
+
+void Transaction::Commit(CommitCallback&& callback) {
+  EnsureCommitNotCalled();
+
+  // If there was an error writing, raise that error now
+  if (!last_write_error_.ok()) {
+    callback(last_write_error_);
+    return;
+  }
+
+  // Make a list of read documents that haven't been written.
+  std::unordered_set<DocumentKey, DocumentKeyHash> unwritten;
+  for (const auto& kv : read_versions_) {
+    unwritten.insert(kv.first);
+  };
+  // For each mutation, note that the doc was written.
+  for (FSTMutation* mutation : mutations_) {
+    unwritten.erase(mutation.key);
+  }
+
+  if (!unwritten.empty()) {
+    // TODO(klimt): This is a temporary restriction, until "verify" is supported
+    // on the backend.
+    callback(
+        Status{FirestoreErrorCode::FailedPrecondition,
+               "Every document read in a transaction must also be written in "
+               "that transaction."});
+  } else {
+    committed_ = true;
+    datastore_->CommitMutations(mutations_, std::move(callback));
+  }
+}
+
+void Transaction::EnsureCommitNotCalled() {
+  HARD_ASSERT(!committed_, "A transaction object cannot be used after its "
+                           "update callback has been invoked.");
+}
+
+absl::optional<SnapshotVersion> Transaction::GetVersion(
+    const DocumentKey& key) const {
+  auto found = read_versions_.find(key);
```

Comment:
This method doesn't exist on other platforms, but makes the code accessing the map more similar (other platforms use some `Get` which returns an element or null).